[PiranhaJava] Remove enum constants with field matching flag value

tnterdan opened the pull request:
Fixes https://github.com/uber/piranha/issues/141

Adds an optional top-level field to the config called `enumProperties`. What this field does, is if the user specifies an enum class name, Piranha will remove enum constants that have a constructor with an argument that matches that `FlagName` value, along with their usages.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/piranha?pullRequest=142) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/piranha?pullRequest=142) before we can accept your contribution.<br/><hr/>**Ty Terdan** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/piranha?pullRequest=142) it.</sub>

mkr-plse reviewed (commented):
Thanks for the PR @tnterdan. Added a few comments. Let us further iterate on `isMatchingEnumFieldValue` implementation as it is invoked for all matching FF API.

tnterdan commented on java/piranha/src/test/java/com/uber/piranha/PiranhaTestingHelpers.java:
```diff
@@ -23,4 +23,38 @@ public static BugCheckerRefactoringTestHelper addHelperClasses(
+            "package com.uber.piranha;",
+            "public enum TestExperimentName {",
+            " STALE_FLAG(\"stale.flag\"),",
+            " OTHER_FLAG(\"other\");",
```
Good point. Will add a TODO, as this should probably be a separate fix since I believe Piranha already doesn't handle removal of the last enum constant with the semicolon properly (it only checks for enums ending with a `,` rather than `,` or `;`)

tnterdan replied:
Created a ticket https://github.com/uber/piranha/issues/143

lazaroclapp commented on java/piranha/src/main/java/com/uber/piranha/config/PiranhaMethodRecord.java:
```diff
@@ -102,14 +87,7 @@ private static String getValueStringFromMap(Map<String, Object> map, String key)
-      }
-    }
-    return null;
+    return PiranhaRecord.getArgumentIndexFromMap(map, ARGUMENT_INDEX_KEY);
```
Again, feel free to just inline this, rather than making a new method just for passing `ARGUMENT_INDEX_KEY` :)

lazaroclapp commented on java/piranha/src/main/java/com/uber/piranha/config/PiranhaRecord.java:
```diff
@@ -0,0 +1,43 @@
+package com.uber.piranha.config;
```
Agreed! (Also, copyright header, please :) )

lazaroclapp commented on java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java:
```diff
@@ -1626,6 +1628,99 @@ public void exceptionWhenArgumentIndexNotSpecifiedAndNotArgumentIndexOptional()
     flagCleaner.init(b.build());
   }
 
+  private BugCheckerRefactoringTestHelper enumUnchangedTest(BugCheckerRefactoringTestHelper bcr) {
```
I still don't quite understand this method or its comment. What is this testing _or_ setting up for other tests?

tnterdan replied:
Good point, the `Test` in the name isn't helping here... I'll think of something clearer. Or it may also be cleaner to keep some duplication in tests for readability's sake.

lazaroclapp commented on java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java:
```diff
@@ -1626,6 +1628,106 @@ public void exceptionWhenArgumentIndexNotSpecifiedAndNotArgumentIndexOptional()
+  }
+
+  /**
+   * [https://github.com/uber/piranha/issues/141]
```
One line description here? (I like the link, but it should be possible to get some idea what this is about without following the link. Also, if we ever move away from GitHub for issue tracking or relocate to another repo/org, some documentation should remain in the code itself)

lazaroclapp replied:
Actually, looking at the paragraph below, this is fine, I think... though maybe I would still do for these tests:

```
Testing enumProperties configuration.

<p>[the test specific paragraph you have written for each test]

<p> See also: [https://github.com/uber/piranha/issues/141]
```

lazaroclapp commented on java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:
```diff
@@ -499,4 +501,201 @@ public void testIgnoresPrefixMatchFlag() throws IOException {
   }
+
+  /**
+   * [https://github.com/uber/piranha/issues/141]
```
Again, I'd move the issue links to the end, start with a line on what is being tested (which can be the existing next line)

lazaroclapp commented on java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:
```diff
@@ -499,4 +501,201 @@ public void testIgnoresPrefixMatchFlag() throws IOException {
+   * <p>Also tests flags that use unqualified enum constant usages, and enum method calls
+   */
+  @Test
+  public void testRemoveEnumFieldValueMatchUntreated() {
```
We "recently" put some overdue work to partition `CorePiranhaTest` into multiple test classes, which is both better for readability and allows running our tests in parallel with gradle.

Could we get all tests related to enum constant support into their own test suite class? (I know that's a small amount of boilerplate, but we really don't want to end up with 2-3k  `CorePiranhaTest` again 😅

lazaroclapp reviewed (changes requested):
Overall this is all really cool work! A few comments/changes requested, though :)

Also, we land PRs by squashing, so if you could change the description to something suitable to be the commit message for this PR, then that would be appreciated.

Again, thanks so much for contributing!

tnterdan commented on java/piranha/src/main/java/com/uber/piranha/config/PiranhaEnumRecord.java:
```diff
@@ -0,0 +1,78 @@
+package com.uber.piranha.config;
```
Added!

tnterdan commented:
Just want to say thanks an absolute ton for the review comments so far @mkr-plse and @lazaroclapp 👏 I know how time consuming code reviews can be, so I super appreciate your attention to detail! 🌟

tnterdan commented on java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:
```diff
@@ -385,10 +392,97 @@ private API getXPAPI(
+      }
+      MemberSelectTree memberTree = (MemberSelectTree) methodTree.getMethodSelect();
+      if (memberTree.getExpression().getKind() == Kind.IDENTIFIER) {
+        // Case 3: Unqualified enum constant with method call (e.g. "STALE_FLAG.getKey()")
```
Hmm so after thinking about this, the issue is if we remove the constant `STALE_FLAG`, it doesn't matter what method call we find, we need to remove all usages because we removed the enum constant itself.

There is a potential use case where the constant stays, and we remove ONLY based on method call. As of right now, that doesn't seem like a useful use case to me. (I could be wrong though!)

Will add a comment to clarify at the very least.

lazaroclapp replied:
Haven't done a full second pass yet, but that's a very valid point. Since the methods are instance methods, we also can't associate them just with the enum, so, removing the enum constant requires removing any calls to methods of that enum constant. Now, the question is, how do we simplify code around that call. Consider the following:

```
public enum ButtonColorFlag {
    // ... a hundred+ other flags
    BLUE_BUTTON_FLAG("button.is.blue", "#0000FF"),
    RED_BUTTON_FLAG("button.is.red", "#FF0000");

    private final String key;
    private final String value;

    public String getKey() {
        return key;
    }

    public String getValue() {
        return value;
    }

    FeatureFlags(final String key, final String value) {
        this.key = key;
        this.value = value;
    }
}
```

Now, presumably code calling `BLUE_BUTTON_FLAG.getKey()` will only happen inside whatever checks for whether the feature is toggled or not (`featureFlagClient.isFeatureEnabled` in #141), but `getValue()` could be called directly, with the value meaning something and, more importantly being assigned or passed to some other field/call:

```
this.button = new Button(color=BLUE_BUTTON_FLAG.getValue())
```

So, what is Piranha to do then?

a) Abort? (e.g. we just don't allow flag enums to carry values others than those needed to check the flag name/identifier)
b) Remove the line? Possibly leaving `this.button` unassigned and breaking code? (In our internal deployment, we don't require Piranha to always produce building code, as some refactorings will require manual follow up no matter what, but we do aim for it)
c) Try to propagate `value`? (Easy-ish in the example above, but impossible in the general case where the enum contains methods that aren't statically evaluatable)

lazaroclapp replied:
Good point. I actually just made a quick test case for it. You are correct, (b) is the current behavior, and this change preserves it, which makes sense to me. Wonder if we could in the future add some logic to Piranha to replace statically evaluatable methods of enum constant with their constant values but, in the limit, this is impossible to handle for all cases of enum methods.

mkr-plse commented on java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:
```diff
@@ -385,10 +392,66 @@ private API getXPAPI(
+  }
+
+  private boolean isMatchingEnumFieldValue(ExpressionTree argTree, VisitorState state) {
+    if (argTree.getKind() != Kind.METHOD_INVOCATION) {
```
Here is another proposal. I believe the enum declarations will be processed by ErrorProne before the invocations are processed. Is it possible to infer the enum flag declarations and methods when the enum class is parsed, store the results and reuse it in `getXPAPI`? 

This will be a problem if the enum declarations are in a different compilation unit (or in a library jar) and are not part of the current compilation. To overcome this, we can have an option to check for the stored results and if unavailable, perform the computation that you have currently, and this option can be turned off by default. 

I think this inference and use will suffice in a majority of the cases.

tnterdan replied:
Isn't this where two passes are needed? Because we can't rely on the enum class being parsed before its usages?

Although I am happy to cache results for multiple reasons.

tnterdan commented on java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:
```diff
@@ -385,10 +392,97 @@ private API getXPAPI(
+      VisitorState state,
+      ImmutableCollection<PiranhaEnumRecord> enumRecords) {
+    ClassTree enumClassTree = ASTHelpers.findClass(classSymbol, state);
+    if (enumClassTree == null) {
```
So far I tried to query members of the symbol for the enum class and was able to get the members. However, it seems that the symbol table doesn't hold the actual `"stale.flag"` String value from the constructor.  At least I don't see it in whatever implementation of the JVM I'm on (a version of AdoptOpenJDK 1.8) in the `com.sun.tools.javac.code` package. But I assume the String value has to be accessible outside of source (such as in the bytecode), so I'll keep investigating on how to get it for other targets/jars.

tnterdan replied:
Kind of thinking about this later, we need the source to be able to remove the enum constant in the first place? Does Piranha already have tests that cover the case where we have an enum in another target/jar, and therefore possibly no access to source code, but we clean up the enum's usages anyway while preserving the enum constant?

I'm really not sure how to get access to the constructor argument values from the symbol table. I don't see it represented anywhere in memory.

To avoid having this block the rest of the PR, may be another ticket to do later, and I can add a TODO if that's OK? At least it will work for projects that have their flags self-contained. I think regardless it would be good to add tests for feature flags across multi-module projects like your example.

tnterdan commented on java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:
```diff
@@ -385,10 +392,97 @@ private API getXPAPI(
+    }
+
+    Symbol sym;
+    if (argTree.getKind() == Kind.IDENTIFIER) {
```
Using a switch statement has some scoping issues where I couldn't reuse declarations of `memberTree` within case blocks and would have to either rename one or both `memberTree` variables, or declare it once in the outer scope. So I'm not sure if that's preferred to using a switch or not.

tnterdan commented on java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:
```diff
@@ -385,10 +392,97 @@ private API getXPAPI(
+        this.config.getEnumRecordsForName(enclosingClass.getSimpleName().toString());
+
+    if (!enumRecords.isEmpty()) {
+      return isEnumSymbol(sym.getSimpleName().toString(), enclosingClass, state, enumRecords);
```
Agreed and added caching :)

tnterdan commented:
OK I believe I have addressed or responded to everything brought up so far 👍 Thank you @mkr-plse and @lazaroclapp, let me know if I missed anything (or if you have any ideas on how to get the constructor argument values out of the symbol table, I tried!) Looking forward to your next pass.

mkr-plse commented on java/README.md:
```diff
@@ -238,3 +273,52 @@ IMPORTANT: Please note that the gradle build script included in that directory a
+
+## Limitations
+
+Currently, Piranha does not distinguish between enum usages for a feature flag to be cleaned up, versus for another, non-feature flag usage.
```
Given Piranha removes the method call, once it is configured, I am not sure whether this merits a discussion in the README. This is expected behavior even for non-enum usages where the method cleanup will not happen if the method is not configured. Can we get rid of this discussion?

tnterdan replied:
Yeah wasn't sure if it was helpful, removed. I purposefully put it in its own commit though so it was easy to remove 😂

mkr-plse reviewed (commented):
Overall, looks good to me except for the update to README.

mkr-plse commented:
> Overall, looks good to me except for the update to README.

Also, `java/piranha/src/test/java/com/uber/piranha/PiranhaTestingHelpers.java` needs to be updated to handle GJF (the build seems to have failed due to this).

tnterdan commented:
OK formatting should be fixed, ran `./gradlew verifyGoogleJavaFormat` locally to verify 👍 Thanks for reviews @mkr-plse and @lazaroclapp :)

mkr-plse reviewed (approved):
LGTM. Thanks for the PR again.

tnterdan commented:
Just waiting on a re-review form @lazaroclapp :)

lazaroclapp commented on java/piranha/src/test/java/com/uber/piranha/EnumConstantTest.java:
```diff
@@ -0,0 +1,228 @@
+ * <p>See: <a href="https://github.com/uber/piranha/issues/141">Issue #141</a>
+ */
+@RunWith(JUnit4.class)
+public class EnumConstantTest {
```
👍  on having this be its own test class!

lazaroclapp reviewed (approved):
Hey @tnterdan , sorry for the delay on my review...

Below are a few comments and one or two very minor nits, but overall, I think this looks ready to land!

Might need to be rebased, though :)

tnterdan commented:
Rebased and squashed already reviewed changes. Thanks for the re-review @lazaroclapp! No worries about it taking time, we all have other priorities to attend to :)

Yeah I agree it would be pretty nice to find the enum constructor without the source, but yeah, I think it's worth following up on that with some better multi-target testing and digging into the symbol table API a bit more. At least it's behind a config option so existing users won't be in for any surprises.

I like to avoid putting too many changes into a single PR anyway 😅 I'm sure this was already a doozy to review.

lazaroclapp commented on java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:
```diff
@@ -385,10 +397,130 @@ private API getXPAPI(
+              + "\n\nIf you are trying to use Piranha to clean up enum flags based on the string"
+              + " argument to their constructor (e.g. using enumProperties), the current"
+              + " implementation will fail to match flag usages on a separate target as that"
+              + " containing the enum source, resulting in partial clean up.");
```
👍  (There is a chance that we discover that in practice erroring out here is too strict and we should print a warning, but I definitely prefer to start with the most "fail-fast" possible solution until we know for sure how often this case shows up or if there is any better way to handle it! :) )

tnterdan replied:
I'm generally in favor of failing fast 😁 Hiding errors tends to cause problems later...

lazaroclapp commented on java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java:
```diff
@@ -0,0 +1,12 @@
+package com.uber.piranha;
```
As silly as that might be for this tiny exception class... can we get a copyright header here?

tnterdan replied:
Oops! This is what I get for copying and pasting from `PiranhaConfigurationException` 😂

lazaroclapp replied:
😅  yup, that one is on us... I need to go over the code and fix those java files that are still missing the header.

lazaroclapp reviewed (changes requested):
One small fix requested regarding the copyright header for the new exception file. Other than that, this is ready to land.

tnterdan commented:
Thanks so much for the reviews @lazaroclapp and @mkr-plse!

lazaroclapp reviewed (approved):
Ready to merge if CI passes.

tnterdan commented:
I'm seeing this message:
> Only those with write access to this repository can merge pull requests.

@lazaroclapp or @mkr-plse could one of you merge on my behalf? Thank you! 🙏

lazaroclapp commented:
> I'm seeing this message:
> 
> > Only those with write access to this repository can merge pull requests.
> 
> @lazaroclapp or @mkr-plse could one of you merge on my behalf? Thank you! 🙏

It is done! 🎉 

Thanks again for the PR and for being patient with our reviews and comments!

mkr-plse commented:
Thanks @tnterdan and @lazaroclapp. Glad to see this land :)
